packages/core/src/app/shipping/ShippingHeader.tsx

```diff
@@ -12,37 +18,64 @@ interface ShippingHeaderProps {
-        <Legend testId="shipping-address-heading">
-            <TranslatedString
-                id={
-                    isMultiShippingMode
-                        ? isGuest
-                            ? 'shipping.multishipping_address_heading_guest'
-                            : 'shipping.multishipping_address_heading'
-                        : 'shipping.shipping_address_heading'
-                }
-            />
-        </Legend>
-
-        {shouldShowMultiShipping && (
-            <a
-                data-test="shipping-mode-toggle"
-                href="#"
-                onClick={preventDefault(onMultiShippingChange)}
-            >
-                <TranslatedString
-                    id={isMultiShippingMode ? 'shipping.ship_to_single' : 'shipping.ship_to_multi'}
-                />
-            </a>
-        )}
-    </div>
-);
-
-export default memo(ShippingHeader);
+    isExtensionEnabled,
+    extensionService,
+}) => {
+    const isRegionInUse = Boolean(
+        isExtensionEnabled &&
+            extensionService.isRegionInUse(ExtensionRegion.ShippingShippingAddressFormBefore),
+    );
+
+    useEffect(() => {
+        if (isRegionInUse) {
+            void extensionService.renderExtension(
+                ExtensionRegionContainer.ShippingShippingAddressFormBefore,
+                ExtensionRegion.ShippingShippingAddressFormBefore,
+            );
+
+            return () => {
+                extensionService.removeListeners(ExtensionRegion.ShippingShippingAddressFormBefore);
+            };
+        }
+    }, [extensionService, isExtensionEnabled, isRegionInUse]);
+
+    return (
+        <>
+            {isRegionInUse && (
+                <div id={ExtensionRegionContainer.ShippingShippingAddressFormBefore} />
+            )}
+            <div className="form-legend-container">
+                <Legend testId="shipping-address-heading">
+                    <TranslatedString
+                        id={
+                            isMultiShippingMode
+                                ? isGuest
+                                    ? 'shipping.multishipping_address_heading_guest'
+                                    : 'shipping.multishipping_address_heading'
+                                : 'shipping.shipping_address_heading'
+                        }
+                    />
+                </Legend>
+
+                {shouldShowMultiShipping && (
+                    <a
+                        data-test="shipping-mode-toggle"
+                        href="#"
+                        onClick={preventDefault(onMultiShippingChange)}
+                    >
+                        <TranslatedString
+                            id={isMultiShippingMode ? 'shipping.ship_to_single' : 'shipping.ship_to_multi'}
+                        />
+                    </a>
+                )}
+            </div>
+        </>
+    );
+}
+
+export default withExtension(memo(ShippingHeader));
```

Comment:
Another follow up improvement, maybe we can avoid using HOC for function components and use the useContext hook or create a custom useExtensions hook similar to useLocale.

Reply:
I built 'useExtensions' before, but we have to use HOC now, so I deleted it to keep usage constant.